### PR TITLE
[TRTLLM-7440][fix] Split `fused_input_embed` to separate out host sync

### PR DESCRIPTION
**Pull request by chang-l**
## Summary by CodeRabbit

- New Features
  - Token filtering for multimodal inputs and an orchestrated embedding-fusion step.
  - Models now forward extra kwargs to the fusion step; callers can supply precomputed indices.

- Tests
  - Added CPU/CUDA unit tests covering token filtering, fusion success/failure, and kwargs precedence.

- Documentation
  - Clarified text/mm token indices usage and host-device synchronization considerations.

- Refactor
  - Skip compilation overhead for embedding path on single-GPU runs.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 1]
```diff
@@ -1130,6 +1131,18 @@ def _preprocess_inputs(self, inputs: Dict[str, Any]):
                         self.previous_kv_lens_offsets_cuda[:num_gen_requests])
         return inputs
 
+    def _prepare_multimodal_indices(self, input_ids: list[int]):
```
Comment (2ez4bz) [inline tensorrt_llm/_torch/pyexecutor/model_engine.py:1134]: I'm sure I'm missing something - why is this now done in the `model_engine` rather than in the `fuse_` function(s)? The PR description suggests it's to avoid unnecessary syncs, but having looked at the code for `fuse_input_embeds` and `filter_mm_token_from_input_ids`, I don't really understand why this would remove the need for those syncs. The pytorch code itself seems to have no assumptions on the device.

Comment (chang-l) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:1134]: Because the model engine’s `prepare_input` runs before each forward pass, some CPU–GPU sync/copy is expected at this stage. However, within the engine’s forward pass itself, we expect purely GPU op.

Previously, the `fused_input_embed` in the forward pass introduced an unintended CPU–GPU sync. This happened because `torch.where()` was used to determine # of txt tokens and mm tokens within `input_ids`, which triggered host-side shape queries in order to preallocate memory for the text and vision embeddings.

File: tensorrt_llm/_torch/models/modeling_multimodal_utils.py [hunk 2]
```diff
@@ -105,6 +105,77 @@ def find_uncached_mm_embeds(
     return sliced_mm_embeds
 
 
+def filter_mm_token_from_input_ids(
```
[the remark this answers is not in the document]
Comment (chang-l) [reply tensorrt_llm/_torch/models/modeling_multimodal_utils.py:108]: Yes, thanks for catching it. I will update it to use both `filter_mm_token_from_input_ids` and `fuse_input_embeds_cuda`.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py (hunk 1 above)
Comment (chang-l) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:1134]: some context might be helpul: 
- torch.where(condition) is identical to torch.nonzero(condition, as_tuple=True) 
- when `input_ids` is on CUDA, torch.nonzero trigger host-dev sync: https://docs.pytorch.org/docs/stable/generated/torch.nonzero.html

Comment (2ez4bz) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:1134]: I see, thanks for the explanation! It might be worth leaving this as code comments for future readers :)

Comment (2ez4bz) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:1134]: For my own benefit:

> some CPU–GPU copy is expected at this stage

are you saying that by having it here, we can piggy back off of existing copies to mask the overhead?

**Review by rakib-hasan — APPROVED** [review]
LGTM.

File: tensorrt_llm/_torch/models/modeling_multimodal_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chang-l) [reply tensorrt_llm/_torch/models/modeling_multimodal_utils.py:108]: updated

**Review by 2ez4bz — APPROVED** [review]
LGTM, but I would personally prefer if the new arguments to `fuse_input_embeds` are made explicit.

File: tensorrt_llm/_torch/models/modeling_multimodal_utils.py [hunk 3]
```diff
@@ -105,48 +105,84 @@ def find_uncached_mm_embeds(
...
+        - If (4) multimodal run, mixed batch of context and generation requests, each context request has a multimodal feature --> return only the fused input_embeds of shape [total length, hidden_dim]. For text tokens, LLM embedding layer has already run.
+    Note:
+        This function may involve host-device synchronization if text_token_indices (cuda tensor) and mm_token_indices (cuda tensor) are not provided from kwargs. See filter_mm_token_from_input_ids for more details.
+    """
+    if len(mm_embeds) == 0:
+        return input_ids, None
+
+    mm_embed = torch.cat(mm_embeds, dim=0)
+
+    if kwargs.get("text_token_indices", None) is not None and kwargs.get(
```
[the remark this answers is not in the document]
Comment (chang-l) [reply tensorrt_llm/_torch/models/modeling_multimodal_utils.py:174]: I agree -- will explicitly list them as optional in the kwargs.

File: tensorrt_llm/_torch/modules/embedding.py [hunk 4]
```diff
@@ -207,10 +207,14 @@ def __init__(
...
-            options={"max-autotune": True},
-            disable=not self.enable_torch_compile_for_embedding)
+        if self.tp_size > 1:
+            embedding_ops_func = torch.compile(
+                pre_comm_embedding_ops,
+                options={"max-autotune": True},
+                disable=not self.enable_torch_compile_for_embedding)
+        else:
+            # Skip torch.compile when TP size is 1 to avoid unnecessary host overhead
+            embedding_ops_func = pre_comm_embedding_ops
```
[the remark this answers is not in the document]
Comment (chang-l) [reply tensorrt_llm/_torch/modules/embedding.py:217]: Even with `disable=True` (as default), it also introduces ~1s host overhead from torch.compile call

**Review by amukkara — APPROVED** [review]

File: tensorrt_llm/_torch/models/modeling_multimodal_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (chang-l) [reply tensorrt_llm/_torch/models/modeling_multimodal_utils.py]: updated

**Review by mikeiovine — APPROVED** [review]
`model_engine.py` changes look good to me

**Review by yechank-nvidia — APPROVED** [review]
LGTM
